From 6ad3d599b9554c8ed3d7c24cdc151e6a7317c5a9 Mon Sep 17 00:00:00 2001
Message-Id: <6ad3d599b9554c8ed3d7c24cdc151e6a7317c5a9.1432378444.git.chang-joon.lee@intel.com>
From: Wang Xiaoming <xiaoming.wang@intel.com>
Date: Thu, 21 May 2015 10:29:28 +0800
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: avoid fb being freed immaturely

If we return early due to event space shortage, we will decrease the fb
refcount, thus causes the fb freed while it is still on fbs list. Later
when we walk through the fbs list, trying to remove the fb, we may get
wrong pointer or wrong fb refcount.

This patch should be squashed into:
    commit 9a7a86c04aa31ad573b59a263ec7719fc0dde88d
    Author: Vidya Srinivas <vidya.srinivas@intel.com>
    Date:   Sat Jul 5 00:31:06 2014 +0530

        FOR_UPSTREAM [VPG]: drm/i915: Sprite flip Notification

Issue: IMINAN-35761, IMINAN-35627
Change-Id: I81d94abd21b41a3330b3ecbcf1d2d7bb085a9538
Signed-off-by: Dongxing Zhang <dongxing.zhang@intel.com>
Signed-off-by: Wang Xiaoming <xiaoming.wang@intel.com>
---
 drivers/gpu/drm/drm_crtc.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/drm_crtc.c b/drivers/gpu/drm/drm_crtc.c
index fb24624..c522643 100644
--- a/drivers/gpu/drm/drm_crtc.c
+++ b/drivers/gpu/drm/drm_crtc.c
@@ -2326,7 +2326,6 @@ int drm_mode_setplane(struct drm_device *dev, void *data,
 
 	if (drm_modeset_lock(&crtc->mutex, NULL))
 		DRM_ERROR("CRTC(%d) lock failed\n", crtc->base.id);
-	old_fb = plane->fb;
 
 	if (plane_req->flags & DRM_MODE_PAGE_FLIP_EVENT) {
 		ret = -ENOMEM;
@@ -2353,6 +2352,7 @@ int drm_mode_setplane(struct drm_device *dev, void *data,
 		e->base.destroy = (void (*) (struct drm_pending_event *)) kfree;
 	}
 
+	old_fb = plane->fb;
 	ret = plane->funcs->update_plane(plane, crtc, fb,
 					 plane_req->crtc_x, plane_req->crtc_y,
 					 plane_req->crtc_w, plane_req->crtc_h,
-- 
1.7.9.5

